[Swift] Test simple cases of @_implementationOnly import

jrose-apple opened the pull request:
More tests to come, but this gets the "don't crash" part in place, and help shake out a bug in the Swift implementation.

Builds on #1139, also depends on a small Swift change that isn't on GitHub yet (to not produce an error when an implementation-only import can't be loaded).

jrose-apple commented:
@adrian-prantl, @dcci, am I doing this correctly? Any test-writing feedback?

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/main_executable/Makefile:
```diff
@@ -0,0 +1,21 @@
+
+SomeLibrary.swiftmodule: SomeLibrary.swift
+	$(MAKE) MAKE_DSYM=NO CC=$(CC) SWIFTC=$(SWIFTC) \
+		SWIFTFLAGS_EXTRAS="$(MOD$*_FLAGS)" \
```
How is `$(MOD$*_FLAGS)` expanded?

jrose-apple replied:
Oops, that's left over from what I copied from. I should simplify this.

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/main_executable/TestMainExecutable.py:
```diff
@@ -0,0 +1,86 @@
+
+    mydir = TestBase.compute_mydir(__file__)
+
+    def launch_and_break(self):
```
This entire function can be replaced by `lldbutil.run_to_source_breakpoint(self, "break here", SBFileSpec("main.swift"), exe_name=self.getBuildArtifact("main"))`

jrose-apple replied:
Ah! Where would I learn about this? Or should I just take a look at everything in lldbutil?

dcci replied:
I don't think it's documented anywhere, but arguably it should.

adrian-prantl replied:
The example test uses it: https://github.com/apple/swift-lldb/blob/stable/packages/Python/lldbsuite/test/sample_test/TestSampleTest.py

I know.. "We have an example test?"

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/main_executable/TestMainExecutable.py:
```diff
@@ -0,0 +1,86 @@
+        self.addTearDownHook(cleanup)
+        self.launch_and_break()
+
+        self.expect("fr var", substrs=[
```
A slightly more robust way of writing this is
```
var = frame.FindVariable("value")
first = var.GetChildMemberWithName("first")
lldbutil.check_variable(self, first, value="2")
...
```

this way you don't depend on LLDB's formatting

jrose-apple replied:
I was deliberately not doing that because it made it harder to observe the structure and gave worse error messages when it failed. I admit to thrashing around on this test case, though.

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/main_executable/TestMainExecutable.py:
```diff
@@ -0,0 +1,86 @@
+        self.expect("fr var", substrs=[
+            "(SomeLibrary.TwoInts) value = (first = 2, second = 3)",
+            "(main.ContainsTwoInts) container = {\n  wrapped = (first = 2, second = 3)\n  other = 10\n}"])
+        self.expect("e value", substrs=["(SomeLibrary.TwoInts)", "= (first = 2, second = 3)"])
```
For expressions, this is as good as it gets, though

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/main_executable/TestMainExecutable.py:
```diff
@@ -0,0 +1,86 @@
+import commands
+import lldb
+from lldbsuite.test.lldbtest import *
+import lldbsuite.test.decorators as decorators
```
We usually `from lldbsuite.test.decorators import *`

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/main_executable/main.swift:
```diff
@@ -0,0 +1,13 @@
+@_implementationOnly import SomeLibrary
```
What does this do?
And perhaps add a comment explaining it?

jrose-apple replied:
This is the new feature being tested. I don't think it makes sense to explain it here; eventually it'll be an actual Swift language feature.

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/main_executable_resilient/TestMainExecutableResilient.py:
```diff
@@ -0,0 +1,86 @@
+    mydir = TestBase.compute_mydir(__file__)
+
+    def launch_and_break(self):
+        exe_name = "main"
```
See above.

jrose-apple replied:
Yeah, this whole directory is identical to the other one but with `-enable-library-evolution` added to the Makefile.

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/main_executable_resilient/TestMainExecutableResilient.py:
```diff
@@ -0,0 +1,86 @@
+        self.addTearDownHook(cleanup)
+        self.launch_and_break()
+
+        self.expect("fr var", substrs=[
```
ditto

adrian-prantl commented:
Look pretty good so far! Could you perhaps create the pull request for swift-5.1-branch instead? We have an automerger from 5.1 -> stable -> upstream-with-swift.

dcci commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/main_executable/TestMainExecutable.py:
```diff
@@ -0,0 +1,86 @@
+        # This one's probably unavoidable, though.
+        self.expect("e TwoInts(4, 5)", error=True, substrs=["error: use of unresolved identifier 'TwoInts'"])
+
+if __name__ == '__main__':
```
You don't really need this `if __name__ == main` and below. Some tests have it, but it's just an historical artifact.

jrose-apple commented:
> Could you perhaps create the pull request for swift-5.1-branch instead? We have an automerger from 5.1 -> stable -> upstream-with-swift.

The Swift change hasn't made it to 5.1 yet (it hasn't even landed in master yet), but sure, I'll do the overall change that way.

dcci commented:
I think the structure of the tests is generally really good. I guess what we have are just minor comments.

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/main_executable_resilient/TestMainExecutableResilient.py:
```diff
@@ -0,0 +1,86 @@
+    
+    @decorators.swiftTest
+    def test_implementation_only_import_main_executable_resilient(self):
+        """Test `@_implementationOnly import` in the main executable with a resilient library"""
```
Could you please just add a sentence or two outlining why `@_implementationOnly` is special and how it is expected to deviate or not from the normal behavior?
Inevitably, at some point somebody else will have to do maintenance on the test and without a comment explaining what behavior needs to be preserved, it can easily happen that the test devolves into something that isn't meaningful any more.

jrose-apple committed:
[Swift] Test simple cases of @_implementationOnly import

More tests to come, but this gets the "don't crash" part in place,
and helped shake out a bug in the Swift implementation.

jrose-apple commented:
Moving to the 5.1 branch.
